features/storage/kvstore/kv_map/kv_map.cpp

```diff
@@ -0,0 +1,216 @@
+        ret =  MBED_ERROR_OUT_OF_MEMORY;
+        goto exit;
+    }
+
+    kv_map_table[kv_num_attached_kvs].partition_name = strdup(partition_name);
+    kv_map_table[kv_num_attached_kvs].kvstore_instance = kv_instance;
+    kv_num_attached_kvs++;
+
+exit:
+    mutex->unlock();
+    return ret;
+}
+
+int kv_detach(const char *partition_name)
+{
+    int ret = MBED_SUCCESS;
+
+    mutex->lock();
+
+    if (!is_initialized) {
+        ret = MBED_ERROR_NOT_READY;
+        goto exit;
+    }
+
+    ret = MBED_ERROR_ITEM_NOT_FOUND;
+    for (int i = 0; i < kv_num_attached_kvs; i++ ) {
+
+        if (strcmp(partition_name, kv_map_table[i].partition_name) != 0) {
+            continue;
+        }
+
+        free(kv_map_table[i].partition_name);
+        kv_map_table[i].kvstore_instance->deinit();
+
+        memcpy(&kv_map_table[i], &kv_map_table[i + 1], sizeof(kv_map_entry_t) * (MAX_ATTACHED_KVS - i - 1));
+        kv_map_table[MAX_ATTACHED_KVS - 1].partition_name = NULL;
+        kv_map_table[MAX_ATTACHED_KVS - 1].kvstore_instance = NULL;
+        kv_num_attached_kvs--;
+        ret = MBED_SUCCESS;
+        break;
+    }
+
+exit:
+    mutex->unlock();
+    return ret;
+}
+
+int kv_deinit()
+{
+    int ret = MBED_SUCCESS;
+
+    mutex->lock();
+
+    if (!is_initialized) {
+        ret = MBED_ERROR_NOT_READY;
+        goto exit;
+    }
+
+    for (int i = 0; i < kv_num_attached_kvs; i++ ) {
+
+        if (kv_map_table[i].kvstore_instance == NULL) {
+            goto exit;
+        }
+
+        free(kv_map_table[i].partition_name);
+        if (kv_map_table[i].kvstore_instance != NULL) {
+            kv_map_table[i].kvstore_instance->deinit();
+        }
+        kv_map_table[i].partition_name = NULL;
+        kv_map_table[i].kvstore_instance = NULL;
+    }
+
+exit:
+    kv_num_attached_kvs = 0;
+    mutex->unlock();
+    return ret;
+}
+
+//return -1 if delimiter is not found
+int kv_get_delimiter_index(const char *str, char delimiter)
+{
+    size_t len = strlen(str);
+    for (size_t i = 0; i < len; i++ ) {
+        if (str[i] != delimiter) {
+            continue;
+        }
+
+        return i;
+    }
+
+    return -1;
+}
+
+// Full name lookup and then break it into KVStore instance and key
+int kv_lookup(const char *full_name, KVStore **kv_instance, size_t *key_index)
+{
+    int ret = MBED_SUCCESS;
+    int delimiter_index;
+    int i;
+    mutex->lock();
+    const char *temp_str = full_name;
+
+    if (!is_initialized) {
+        ret = MBED_ERROR_NOT_READY;
+        goto exit;
+    }
+
+    *key_index = 0;
+    if (*temp_str == '/') {
+        temp_str++;
+        (*key_index)++;
+    }
+
+    delimiter_index = kv_get_delimiter_index(temp_str, '/');
+    if (delimiter_index == -1) { //delimiter not found
+        *kv_instance = kv_map_table[0].kvstore_instance;
+        goto exit;
+    }
+
+    for (i = 0; i < kv_num_attached_kvs; i++ ) {
+
+        if (memcmp(temp_str, kv_map_table[i].partition_name, delimiter_index) != 0) {
```

Comment:
note that if partition_name is shorter than delimiter_index memcmp might crash on access violation. 
Better use strcmp